DNS Cloud Services Provider

gambol99 opened the pull request:
## Summary

Adding a serviceprovider to service dns hosting domains in route53. This also deploy the workload identity role and external-dns services in the cluster to consume them.

## Checklist

 - [ ] I've created a PR to update the [documentation](https://github.com/appvia/kore-docs): PR LINK

## Changes

Adds a serviceprovider used to provision hosting domains for route53 (gcp coming). The service also creates the workload identity and deploys the in-cluster resources for users to consume them

## Testing

- Create an EKS cluster
- Provision a Route53 Domain
- Use via a service in the cluster

## Follow-up stories

- We need to add reclainPolicy to the services to allow for cloud resources to live on beyond deletion
- We need to remove the external-dns components into operators

mrsheepuk commented on pkg/serviceproviders/domains/plans.go:
```diff
@@ -0,0 +1,199 @@
+				},
+				Spec: servicesv1.ServicePlanSpec{
+					Kind:        GCPDomainServiceKindName,
+					Summary:     "Creates a publically facing domain in GCP for your applications",
```
`public-facing` instead of `publically facing`

gambol99 replied:
done

mrsheepuk commented on pkg/serviceproviders/domains/route53.go:
```diff
@@ -0,0 +1,485 @@
+			return reconcile.Result{}, err
+		}
+
+		session := awsutils.GetSession(creds, r.eks.Spec.Region)
```
Design-wise, I don't like the above - I think that the service provider should take a dependency on an interface which says 'give me an AWS session for this region' then whatever implementation of that interface is provided should deal with adding the relevant credentials to it if needed - then, it puts the logic in one place that we'd need to touch if we wanted to move to using workload identity or IAM roles or whatever.

Each time we're adding in this logic about having essentially a hard-coded set of credentials is another place we'll need to change things when we come to sort out IAM roles/permissions correctly for everything we're doing.

I've implemented [something like that for AWS in kore-costs](https://github.com/appvia/kore-costs/blob/aws-cost-provider/pkg/providers/aws/awsclient.go) (and [an implementation of the session source](https://github.com/appvia/kore-costs/blob/aws-cost-provider/pkg/providers/aws/config.go) ) - can we use something similar here? It can be a separate PR to refine this everywhere, but I think we should consider it before building loads more providers using the hard-coded creds concept.

mrsheepuk commented on pkg/serviceproviders/domains/route53.go:
```diff
@@ -0,0 +1,485 @@
+			return reconcile.Result{}, err
+		}
+
+		// @TODO add the tags for costs
```
this is a one-liner at this point assuming CreateHostedZoneInput has tags on it, particularly as you already have the references to the owning cluster, so we should do it.

gambol99 replied:
added

mrsheepuk commented on pkg/serviceproviders/domains/route53.go:
```diff
@@ -0,0 +1,485 @@
+		config.Policy = Route53IAMPolicy
+		config.RoleDescription = fmt.Sprintf("Role for external-dns: %s/%s", r.cluster.Namespace, r.cluster.Name)
+		config.RoleName = r.WorkloadIdentityRoleName(r.cluster)
+		config.ServiceAccountNamespace = "kore-operators"
```
feels like 'kore-operators' should be in a constant somewhere 

EDIT: it looks like you have got a constant for it used just below - kore.HubOperators - use that instead of the hard-coded string?

gambol99 replied:
done

gambol99 commented on pkg/serviceproviders/domains/plans.go:
```diff
@@ -0,0 +1,199 @@
+					Schema:      string(jsonutils.MustCompact([]byte(Route53Kind))),
+				},
+			},
+			{
```
I'll drop the servicekind for now until implemented ..

mrsheepuk commented on pkg/serviceproviders/domains/route53.go:
```diff
@@ -0,0 +1,485 @@
+			Helm: &application.HelmAppV1Helm{
+				Name:    "external-dns",
+				URL:     "https://charts.bitnami.com/bitnami/",
+				Version: "3.3.0",
```
i'm uneasy about the version number being hard-coded this deep inside our code, though i don't have a suggestion where it should be... feels like somehow this service provider should simply be able to say 'give me the current kore version of this helm chart'. Perhaps another one to consider later but before we create too much spider, we should think about this.

gambol99 replied:
I'll be moving this logic out at some point and these plans should be populate by the catalog ..

mrsheepuk commented on pkg/serviceproviders/domains/route53.go:
```diff
@@ -0,0 +1,485 @@
+			"Policy":    "upsert-only",
+			"Region":    r.eks.Spec.Region,
+			"Role":      r.WorkloadIdentityRoleName(r.cluster),
+		}
```
edge-case but I assume from the shape of this code that it's not possible to use Route53 external DNS with a GKE cluster?

gambol99 replied:
At the moment no as services as tied to clusters

mrsheepuk commented on pkg/serviceproviders/domains/route53_delete.go:
```diff
@@ -0,0 +1,235 @@
+			// if nothing to delete we can skip
+			if len(batch.Changes) == 0 {
+				continue
+			}
```
conceptually, should we not be deleting only the records we've created, rather than all of them?

gambol99 replied:
We can add as an iteration ... it would be fiddly to do so, as the records as populated by services in a kubernetes clusters not kore directly. I believe external-dns add text records for updates ..so we've have to parse the records, found matches A / CNAMES and remove those .. and anything else left over we'd have to throw and error and ask them to remove manually i guess (or by whatever means they were added)

gambol99 commented:
I've fixed the following from the review
* 7aba7cc8 - we will only delete the hosting IF we created it for now (10 minutes ago) <Rohith>
* df19dbc6  - adding the tags for costs for the created zone (11 minutes ago) <Rohith>
* 76e05ee9 - fixing the typo and removed the gcp kind from being enabled for now (11 minutes ago) <Rohith>

I've kept the credentials thing for now as the change would be quite large (time pressure) .. but also because the intent is to remove the requirement for a service to tie to a cluster thus would probably have to change it twice ..

mrsheepuk commented on pkg/serviceproviders/domains/route53.go:
```diff
@@ -227,13 +227,30 @@ func (r *route53Impl) EnsureHostingZone() controllers.EnsureFunc {
+			AddTags: []*route53.Tag{
+				{Key: aws.String(kore.Label("owned")), Value: aws.String("true")},
+				{Key: aws.String(kore.Label("team")), Value: aws.String(teamID)},
+				{Key: aws.String(kore.Label("cluster")), Value: aws.String(clusterID)},
```
two things to fix up here:

1. Need the instance as well so kore knows that /this/ kore owns this thing... 
2. The tag names are not kore.Labels for costs (to be identical across cloud providers they cannot contain . or / characters)... see pkg/clusterproviders/tagging.go for the reference.

mrsheepuk commented:
> I've kept the credentials thing for now as the change would be quite large (time pressure) .. but also because the intent is to remove the requirement for a service to tie to a cluster thus would probably have to change it twice ..

:+1: Just wanted to open the discussion about that, we can handle it separately but we should start thinking about it.

gambol99 commented:
Fixed the tagging [here](https://github.com/appvia/kore/pull/1269/commits/1d4a7e9ed9f775c954ed176f48f4f006f3c588f4) and tested as working ..

mrsheepuk commented on pkg/serviceproviders/domains/route53_delete.go:
```diff
@@ -0,0 +1,269 @@
+					continue
+				}
+
+				return true
```
This seems like a very long-winded way to write:

```
if aws.StringValue(x.Key) == "kore-instance" {
    return aws.StringValue(x.Value) == labels[kore.LabelKoreIdentifier]
}
```

... but if it all works and you're happy, happy to approve as-is.

gambol99 replied:
Far better ... :+1:

gambol99 replied:
https://github.com/appvia/kore/pull/1269/commits/88131718e5def6e15f836ab241fcf71d8dc2c0ff

mrsheepuk reviewed (approved):
LGTM with the changes made
